include/triggeralgs/TriggerActivityFactory.hpp

```diff
@@ -0,0 +1,43 @@
+/* @file: TriggerActivityFactory.hpp
+ *
+ * This is part of the DUNE DAQ Application Framework, copyright 2023.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef TRIGGERALGS_TRIGGER_ACTIVITY_FACTORY_HPP_
+#define TRIGGERALGS_TRIGGER_ACTIVITY_FACTORY_HPP_
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include "triggeralgs/TriggerActivityMaker.hpp"
+
+#define REGISTER_TRIGGER_ACTIVITY_MAKER(tam_name, tam_class) \
+  bool tam_class::s_registered = TriggerActivityFactory::registerCreator(tam_name, []()->std::shared_ptr<TriggerActivityMaker> {return std::make_shared<tam_class>();});
+
+namespace triggeralgs {
+
+class TriggerActivityFactory
+{
+  public:
+    using TAMakerCreator = std::shared_ptr<TriggerActivityMaker>(*)();
+    using TAMakerMap = std::unordered_map<std::string, TAMakerCreator>;
+
+  public:
+    TriggerActivityFactory(const TriggerActivityFactory&) = delete;
+    TriggerActivityFactory& operator=(const TriggerActivityFactory&) = delete;
+    virtual ~TriggerActivityFactory() {}
+
+    static std::shared_ptr<TriggerActivityMaker> makeTAMaker(const std::string& algName);
```

Comment:
Why static, and why shared pointer rather than unique?

We might want multiple instances of the same algorithm (unsure why in reality, but for testing we could use this feature. E.g. in the last coldbox we were going to look at how using multiple algorithms scales with the number of them by adding Prescale TAMs).

TAMs are also not shared between different objects, so unique pointer would make more sense here.

Reply:
The shared pointer was just keeping up with the previous implementation. It didn't feel necessarily "wrong," so I kept it that way. If a unique pointer is better, then I agree it should be changed.

I may be naive: does this static not apply to the function makeTAMaker? The returned shared pointers would not be static.